dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/command/handler/ReRunWorkflowCommandHandler.java

```diff
@@ -0,0 +1,102 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.server.master.engine.command.handler;
+
+import org.apache.dolphinscheduler.common.enums.CommandType;
+import org.apache.dolphinscheduler.common.enums.WorkflowExecutionStatus;
+import org.apache.dolphinscheduler.dao.entity.Command;
+import org.apache.dolphinscheduler.dao.entity.ProcessInstance;
+import org.apache.dolphinscheduler.dao.entity.TaskInstance;
+import org.apache.dolphinscheduler.dao.repository.ProcessInstanceDao;
+import org.apache.dolphinscheduler.dao.repository.TaskInstanceDao;
+import org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteContext.WorkflowExecuteContextBuilder;
+
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.stereotype.Component;
+
+/**
+ * This handler used to handle {@link CommandType#REPEAT_RUNNING} which will rerun the workflow instance.
+ */
+@Component
+public class ReRunWorkflowCommandHandler extends RunWorkflowCommandHandler {
+
+    @Autowired
+    private ProcessInstanceDao workflowInstanceDao;
+
+    @Autowired
+    private TaskInstanceDao taskInstanceDao;
+
+    @Autowired
+    private ApplicationContext applicationContext;
+
+    /**
+     * Generate the repeat running workflow instance.
+     * <p> Will use the origin workflow instance, but will update the following fields. Need to note we cannot not
+     * update the command params here, since this will make the origin command params lost.
+     * <ul>
+     *     <li>state</li>
+     *     <li>command type</li>
+     *     <li>start time</li>
+     *     <li>restart time</li>
+     *     <li>end time</li>
+     *     <li>run times</li>
+     * </ul>
+     */
+    @Override
+    protected void assembleWorkflowInstance(final WorkflowExecuteContextBuilder workflowExecuteContextBuilder) {
+        final Command command = workflowExecuteContextBuilder.getCommand();
+        final int workflowInstanceId = command.getProcessInstanceId();
+        final ProcessInstance workflowInstance = workflowInstanceDao.queryOptionalById(workflowInstanceId)
+                .orElseThrow(() -> new IllegalArgumentException("Cannot find WorkflowInstance:" + workflowInstanceId));
+        workflowInstance.setVarPool(null);
+        workflowInstance.setStateWithDesc(WorkflowExecutionStatus.RUNNING_EXECUTION, command.getCommandType().name());
+        workflowInstance.setCommandType(command.getCommandType());
+        workflowInstance.setStartTime(new Date());
```

Comment:
should not update start time when repeat running?

Reply:
Yes, removed.